### PR TITLE
Add gnmi test for smartswitch

### PR DESCRIPTION
**Pull request by ganglyu**
### Description of PR

Summary:
Fixes # (issue)
Microsoft ADO: 28940620

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405

#### What is the motivation for this PR?
Add gnmi end to end test for smartswitch

#### How did you do it?
Use GNMI to update the DASH_VNET_TABLE, and use redis cli to verify the DASH_VNET_TABLE in APPL_DB.
Only run test for smartswitch device.

#### How did you verify/test it?
Run gnmi end to end test

### CODE REVIEW

File: tests/gnmi/helper.py [hunk 1]
```diff
@@ -96,6 +96,7 @@ def apply_cert_config(duthost):
     dut_command += "--config_table_name GNMI_CLIENT_CERT "
     dut_command += "--client_auth cert "
     dut_command += "--enable_crl=true "
+    dut_command += "--zmq_address=tcp://127.0.0.1:8100 "
```
Comment (qiluo-msft) [inline tests/gnmi/helper.py:99]: >zmq_address [](http://example.com/codeflow?start=22&length=11)

Is it possible that this function will be on a platform without zmq? If yes, we need to conditionally use `--zmq_address` parameter.

File: tests/gnmi/test_gnmi_smartswitch.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    logger.info("type {}, subtype {}".format(type, subtype))
+    if type != "LeafRouter" or subtype != 'SmartSwitch':
+        pytest.skip("This test is supported only on smartswitch platforms")
+    # Locate the first online DPU
+    # DPU0            N/A              N/A         Online              up       N/A
+    target = None
+    result = duthost.shell("show chassis module status")
+    for line in result['stdout_lines']:
+        module_status = line.split()
+        if "Online" in module_status:
```
Comment (qiluo-msft) [inline tests/gnmi/test_gnmi_smartswitch.py:42]: >module_status [](http://example.com/codeflow?start=23&length=13)

This check is risky. Suggest
1. parse the table header line to get all the columns' names
2. get the index of expected column
3. loop all the remaining lines, check the expected column which have exact match of "Online"

File: tests/gnmi/test_gnmi_smartswitch.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            logger.info("target is {}".format(target))
+            break
+    assert target is not None, "Can't locate online DPU"
+    # Get redis port
+    result = duthost.shell("cat /var/run/redis%s/sonic-db/database_config.json" % target)
+    data = json.loads(result['stdout'])
+    redis_port = data['INSTANCES']['redis']['port']
+    file_name = "vnet.txt"
+    vni = "1000"
+    guid = "559c6ce8-26ab-4193-b946-ccc6e8f930b2"
```
Comment (qiluo-msft) [inline tests/gnmi/test_gnmi_smartswitch.py:53]: >559c6ce8 [](http://example.com/codeflow?start=12&length=8)

How did you get this guid string?

File: tests/gnmi/helper.py (hunk 1 above)
Comment (ganglyu) [reply tests/gnmi/helper.py]: Updated.

File: tests/gnmi/test_gnmi_smartswitch.py (hunk 2 above)
Comment (ganglyu) [reply tests/gnmi/test_gnmi_smartswitch.py]: Updated.

File: tests/gnmi/test_gnmi_smartswitch.py (hunk 3 above)
Comment (ganglyu) [reply tests/gnmi/test_gnmi_smartswitch.py]: Now I use uuid to generate guid.

File: tests/gnmi/test_gnmi_smartswitch.py (hunk 2 above)
Comment (qiluo-msft) [reply tests/gnmi/test_gnmi_smartswitch.py]: Please initialize name_idx, oper_status_idx to something (like NULL) before the loop, and check valid values after the loop.

I understand the python code is working, but not robust, and easy to trigger bug, if there is global variable with the same names.

Comment (ganglyu) [reply tests/gnmi/test_gnmi_smartswitch.py]: Fixed.

**Review by wen587 — APPROVED** [review]

**Review by wangxin — APPROVED** [review]

**Pull request merged**